pkg/clustermesh/remote_cluster.go

```diff
@@ -113,6 +117,12 @@ func (rc *remoteCluster) getLogger() *logrus.Entry {
 // releaseOldConnection releases the etcd connection to a remote cluster
 func (rc *remoteCluster) releaseOldConnection() {
 	rc.mutex.Lock()
+
+	// cancel the old context
+	if rc.cancelFunc != nil {
+		rc.cancelFunc()
+	}
```

Comment:
Good point, but imagine the etcd clients doesn't establish the connection properly.. I.e. the errorChan never gets closed https://github.com/cilium/cilium/blob/master/pkg/kvstore/etcd.go#L749 and the whole thing times out on this line https://github.com/cilium/cilium/blob/master/pkg/kvstore/etcd.go#L754 
So admin has to wait 15min for the connection to change which is not very cool.

Reply:
So that also means that the `DoFunc` never returns during these 15mins (the default timeout) and during that time there is no way to cleanup the connection and if `rc.releaseOldConnection()` is inside the `DoFunc` that cannot be accessed as well

Reply:
You are right, I overlooked that.

I had again a fresh look at the code, and here are my thoughts:
* I agree with you about the need for the context to be canceled when the controller gets updated, to ensure that the `DoFunc` terminates as soon as possible.
* Still, I wonder if it could make sense to add the context canceling logic to the controller implementation, to be enabled by a new `ControllerParam` (e.g.,` CancelDoFuncOnUpdate`) to preserve the existing behavior in the other cases. I assume it could be a rather general behavior to desire that the `DoFunc` terminates as soon as possible if the controller gets updated. IMHO, it could be handled as part of the`updateParamsLocked` function: https://github.com/cilium/cilium/blob/f9d290c20ee9349da28bf8164ab8e7eb178079b0/pkg/controller/controller.go#L320-L328
* `rc.releaseOldConnection()` should be moved back as the first instruction of the `DoFunc`. Otherwise, you might have a race condition if `restartRemoteConnection` gets called when the `DoFunc` is e.g., somewhere here https://github.com/cilium/cilium/blob/f9d290c20ee9349da28bf8164ab8e7eb178079b0/pkg/clustermesh/remote_cluster.go#L247-L248 (but also a bit earlier, depending on the effect of closing the context on the returned errors). In that case, the struct fields would still be nil and thus not cleaned by `releaseOldConnection()`, leading to a leak. It could also cause other problems since they would be almost immediately set to a non-nil value when the `DoFunc` terminates, while they would be expected to be nil since another run of the DoFunc is then being performed. Instead, if it is part of the `DoFunc`, there is the guarantee that the previous run terminated when a new one is performed and triggers the cleanup.


Does that make sense to you?

Reply:
yes totally! sounds great.. should I try to do that? or do you want to do it yourself? :)

Reply:
Feel free to go ahead if you'd like

Reply:
done